### PR TITLE
Inline refactoring: move state machine into LegacyPolicy

### PR DESCRIPTION
**Pull request by AndyAyersMS**
The state machine used to estimate callee native code size is now an
internal part of the LegacyPolicy.

During `fgFindJumpTargets` the policy determines if the state machine
is need, and if so, sets one up. `fgFindJumpTargets` then notifies the
policy of the various opcodes in the method, and the policy uses this
to drive the state machine forward.

When the IL scan is completed, the `fgFindJumpTargets` notifies the
policy, then, for the discretionary inline case, grabs the native size
estimate, and uses that to evaluate inlinability.

The knowledge of when inlines are "always", "forced", or
"discretionary" is now encapsulated by the policy instead of being
distributed in various places in the code.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@kyulee1 PTAL
/cc @dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
Fixed a clang warning about member initialization order.

**Comment by AndyAyersMS** [comment]
Ubuntu CscBench failing with
```
CscBench.exe Unhandled Exception: System.TypeInitializationException: The type initializer for 'SHA1' threw an exception. ---> System.ArgumentNullException: Value cannot be null. Parameter name: type at System.Reflection.IntrospectionExtensions.GetTypeInfo(Type type) at Roslyn.Utilities.PortableShim.SHA1..cctor() 
```
seems to be hitting a lot of recent PRs.

**Comment by AndyAyersMS** [comment]
@kyulee1 pinging you --  not sure if you saw this one or not...

File: src/jit/flowgraph.cpp [hunk 1]
```diff
@@ -4182,90 +4182,53 @@ class fgStack
     unsigned slot1;
     unsigned depth;
 };
+//------------------------------------------------------------------------
+// fgFindJumpTargets: walk the IL stream, determining jump target offsets
+//
+// Arguments:
+//    codeAddr   - base address of the IL code buffer
+//    codeSize   - number of bytes in the IL code buffer
+//    jumpTarget - [OUT] xxx yuuu
```
Comment (kyulee1) [inline src/jit/flowgraph.cpp]: I don't understand the comment.

File: src/jit/inlinepolicy.cpp [hunk 2]
```diff
@@ -103,6 +103,39 @@ void LegacyPolicy::noteBool(InlineObservation obs, bool value)
...
+        case InlineObservation::CALLEE_BEGIN_OPCODE_SCAN:
+            {
+                // Set up the state machine, if this inline is
+                // discretionary and is still a candidate.
+                if (inlDecisionIsCandidate(inlDecision)
+                    && (inlObservation == InlineObservation::CALLEE_IS_DISCRETIONARY_INLINE))
+                {
+                    // Better not have a state machine already.
+                    assert(inlStateMachine == nullptr);
+                    inlStateMachine = new (inlCompiler, CMK_Inlining) CodeSeqSM;
```
Comment (kyulee1) [inline src/jit/inlinepolicy.cpp]: It appears previously we just get the address from local or object's instance field, but now you create one from the heap. Do we have concern about repeating this step?

**Comment by kyulee1** [comment]
LGTM other than a question/comment.

File: src/jit/flowgraph.cpp (hunk 1 above)
Comment (AndyAyersMS) [inline src/jit/flowgraph.cpp]: Thanks for catching this, I meant to go back and fix it.

Comment (AndyAyersMS) [inline src/jit/flowgraph.cpp]: I'll update this in my next set of changes, if that's ok.

File: src/jit/inlinepolicy.cpp (hunk 2 above)
Comment (AndyAyersMS) [inline src/jit/inlinepolicy.cpp]: Maybe a little, but inline candidates aren't all that common. I'll take a quick look at the memory used during crossgen just to be sure this doesn't add up to anything.

**Comment by kyulee1** [comment]
Okay. All sounds good to me.

**Comment by AndyAyersMS** [comment]
Worst case per-method memory increase when crossgenning mscorlib is around 5K bytes (out of ~780K total), when compiling `System.Internal:CommonlyUsedGenericInstantiations` so a bit less than 1%. The jit attempts 80 discretionary inlines, 60 fail and 20 succeed.

Average case is around 0.1%.

Memory impact on CHK is a bit worse since the state machine drags around a 500 byte table, but still around 1% on average.

This is actually measured on my next change, which includes all the above. I'll make a note to go back and reassess memory footprint when I finish the last bits of legacy policy extraction.

**Comment by briansull** [comment]
LGTM

**Pull request merged**